infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/BaseSnapshotReader.java

```diff
@@ -0,0 +1,326 @@
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.LogReplication;
+import org.corfudb.runtime.exceptions.TrimmedException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.StreamOptions;
+import org.corfudb.runtime.view.stream.OpaqueStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Stream;
+
+import static com.google.protobuf.UnsafeByteOperations.unsafeWrap;
+import static org.corfudb.infrastructure.logreplication.config.LogReplicationConfig.DEFAULT_MAX_DATA_MSG_SIZE;
+import static org.corfudb.protocols.CorfuProtocolCommon.getUuidMsg;
+import static org.corfudb.protocols.service.CorfuProtocolLogReplication.generatePayload;
+import static org.corfudb.protocols.service.CorfuProtocolLogReplication.getLrEntryMsg;
+
+/**
+ * Base class for snapshot readers. It provides common functionality for snapshot readers in different
+ * replication models.
+ */
+@Slf4j
+public abstract class BaseSnapshotReader extends SnapshotReader {
+    /**
+     * The max size of data for SMR entries in data message.
+     */
+    private final int maxDataSizePerMsg;
+    private final Optional<DistributionSummary> messageSizeDistributionSummary;
+    private final CorfuRuntime rt;
+    private long snapshotTimestamp;
+    protected Set<String> streams;
+    private PriorityQueue<String> streamsToSend;
+    private long preMsgTs;
+    private long currentMsgTs;
+    private OpaqueStreamIterator currentStreamInfo;
+    private long sequence;
+    private OpaqueEntry lastEntry = null;
+
+    @Getter
+    private ObservableValue<Integer> observeBiggerMsg = new ObservableValue(0);
+
+    protected final LogReplication.LogReplicationSession session;
+    protected final LogReplicationContext replicationContext;
+
+    /**
+     * Init runtime and streams to read
+     */
+    public BaseSnapshotReader(CorfuRuntime runtime, LogReplication.LogReplicationSession session,
+                              LogReplicationContext replicationContext) {
+        this.rt = runtime;
+        this.session = session;
+        this.replicationContext = replicationContext;
+        this.rt.parseConfigurationString(runtime.getLayoutServers().get(0)).connect();
+        this.maxDataSizePerMsg = replicationContext.getConfig(session).getMaxDataSizePerMsg();
+        this.messageSizeDistributionSummary = configureMessageSizeDistributionSummary();
+        refreshStreamsToReplicateSet();
+        log.info("Total of {} streams to replicate at initialization. Streams to replicate={}, Session={}",
+                this.streams.size(), replicationContext.getConfig(session).getStreamsToReplicate(),
+                TextFormat.shortDebugString(session));
+    }
+
+    protected abstract void refreshStreamsToReplicateSet();
```

Comment:
implementation of this method is the same for both supported models.  It can be kept here in the base class

Reply:
Tend to leave it here for now as in next PR it might be different, and for RoutingQueue model I guess it's also needed